### PR TITLE
fix calculation of tuplet baseLen for complex cases

### PR DESCRIPTION
**Pull request by AntonioBL**

### CODE REVIEW

**Comment by AntonioBL** [comment]
I added the check if the denominator of the baseLen is a power of two to prevent an assert failure introduced while converting the fraction to a duration, which is present only in master (reproducible for example by selecting a 4/4 full measure rest and trying to add a 4:5 tuplet inside this duration).

File: mscore/tupletdialog.cpp [hunk 1]
```diff
@@ -103,15 +103,20 @@ Tuplet* MuseScore::tupletDialog()
...
 
-      tuplet->setBaseLen(Fraction(1, f.denominator()));
+      // check if the duration is exactly divisible by the number of actual notes corresponding to the tuplet
+      // and that the result is a valid duration type (1 over a power of two)
+      if ((f.numerator() == 1) && ((f.denominator() & (f.denominator() - 1)) == 0))
+            tuplet->setBaseLen(Fraction(1, f.denominator()));
+      else
+            tuplet->setBaseLen(TDuration::DurationType::V_INVALID);
 
       if (tuplet->baseLen() == TDuration::DurationType::V_INVALID) {
```
Comment (ajyoon) [inline mscore/tupletdialog.cpp]: Why not bring this block directly into the `else` block and skip the `baseLen` setting entirely?

**Comment by lasconic** [comment]
I believe it would be better to have a static function in TDuration to check on the "validity" of a fraction.

**Comment by AntonioBL** [comment]
@ajyoon : I thought the same, but maybe there are corner cases in which the duration is a valid theoretical duration (e.g. 1/1024 or similar), but it does not correspond to a valid duration defined inside MuseScore.

@lasconic : I think there is a check here: https://github.com/musescore/MuseScore/blob/master/libmscore/durationtype.cpp#L405 but this is just quitting the program when the duration is not valid. In the example I was quoting above, the fraction is 1/5 and it is compared to 3/16 inside the assert.

The baselen fix is actually completely independent from this sanity check (it is just the modification at line 106).

**Comment by lasconic** [comment]
I understand. My point was that before the ASSERT was added by @shoogle the fraction validity was tested by creating a TDuration and then checking the validity of it. It's not longer possible since the new TDuration constructor always creates valid duration or asserts. So I believe it would be helpful to have a static function to test the validity of a duration before creating it...

Regarding the original fix, line 206, which bug does it solve? Any tests?

**Comment by AntonioBL** [comment]
@lasconic : yes, I agree. I don't know if I could be able to create such a function.
I think valid durations should have as a denominator a power of two, and as numerator 1, 3 (1+2, e.g. dotted quarter = 3/8), 7 (1+2+4, e.g. double dotted quarter = 7/16), 15 (1+2+4+8, e.g. triple dotted quarter = 15/32), (31 ... ?). Do you think anything else would be a valid duration?

The fix (which is line 106, I mistyped the number) is for https://musescore.org/en/node/136406
Can the tests cover the usage of the tupletdialog (which requires the graphical interface), since this bug can appear only in this case?

**Comment by lasconic** [comment]
For https://musescore.org/en/node/136406, I think I have a fix and a test in https://github.com/musescore/MuseScore/pull/2881 (at least when the tests will pass...)

For the static validity check function, I would do something similar than the TDuration constructor, but instead of assert return false.

**Comment by AntonioBL** [comment]
It came to my mind that full measure rests can also have other durations, for example 5/4, or whatever "strange" numerator we could have. As you point out in the issue report, the use of full measure rests is at the moment the only option to write tuplets spanning over durations different from single note durations.

**Comment by AntonioBL** [comment]
I think that reducing the ratio and taking the denominator is not always correct. It was something I found when trying to reproduce this: https://musescore.org/en/node/137281 see the "moreover" note I added.
When the tuplet was read back from the saved file, its duration was evaluated wrongly and cutting and pasting it was giving problems (on top of the corruption of that bug).

**Comment by lasconic** [comment]
I integrated this fix in https://github.com/musescore/MuseScore/pull/2881 together with some tests. 
The tuplet reading code is now in read114 and read206 for old files which were created with wrong baseLen. I used the trick I previously used but if @AntonioBL you want to do a follow up PR with a better tests for odd tuplets like it's done in MusicXML, it's very much welcome! Also I will wait for your "looks good to merge" on https://github.com/musescore/MuseScore/pull/2881 before pulling it.

**Pull request closed without merging**